// @loaders.gl/polyfills/dist/es6/fetch-node/fetch.node.js
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import fs from 'fs';
import http from 'http';
import https from 'https';
import util from 'util';
import { toArrayBuffer } from './utils/to-array-buffer.node';
import decodeDataUri from './utils/decode-data-uri.node';
import { concatenateReadStream } from './utils/stream-utils.node';
import Headers from './headers.node';
const DEFAULT_OPTIONS = {
  dataType: 'arraybuffer'
};

const isDataURL = url => url.startsWith('data:');

const isRequestURL = url => url.startsWith('http:') || url.startsWith('https:');

class NodeFetchResponse {
  constructor(url, options) {
    this.url = url;
    this.options = options;
    this.bodyUsed = false;
    this._headers = null;
  }

  get headers() {
    this._headers = this._headers || this._getHeaders();
  }

  get body() {
    const {
      url,
      options
    } = this;
    this.bodyUsed = true;
    return createReadStream(url, options);
  }

  arrayBuffer() {
    var _this = this;

    return _asyncToGenerator(function* () {
      _this.bodyUsed = true;
      return readFile(_this.url, _this.options);
    })();
  }

  text() {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      const arrayBuffer = yield _this2.arrayBuffer();
      const textDecoder = new TextDecoder();
      return textDecoder.decode(arrayBuffer);
    })();
  }

  json() {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      const text = yield _this3.text();
      return JSON.parse(text);
    })();
  }

  _getHeaders() {
    const headers = {};

    if (this.httpResponse) {
      const httpHeaders = this.httpResponse.getHeaders();

      for (const name in httpHeaders) {
        const header = headers[name];
        headers[name] = String(header);
      }
    } else {
      const contentLength = this._getContentLength();

      if (Number.isFinite(contentLength)) {
        headers['Content-Length'] = contentLength;
      }
    }

    return new Headers(headers);
  }

  _getContentLength() {
    if (isRequestURL(this.url)) {
      return null;
    } else if (isDataURL(this.url)) {
      return this.url.length - 'data://'.length;
    }

    const stats = fs.statSync(this.url);
    return stats.size;
  }

}

export default function fetchNode(url, options) {
  return new NodeFetchResponse(url, options);
}

function getReadFileOptions() {
  let options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  options = Object.assign({}, DEFAULT_OPTIONS, options);
  options.responseType = options.responseType || options.dataType;

  if (fs) {
    options.encoding = options.encoding || (options.dataType === 'text' ? 'utf8' : null);
  }

  return options;
}

function readFile(_x) {
  return _readFile.apply(this, arguments);
}

function _readFile() {
  _readFile = _asyncToGenerator(function* (url) {
    let options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    options = getReadFileOptions(options);

    if (isDataURL(url)) {
      return Promise.resolve(decodeDataUri(url));
    }

    if (isRequestURL(url)) {
      return new Promise((resolve, reject) => {
        options = _objectSpread({}, new URL(url), options);
        const request = url.startsWith('https:') ? https.request : http.request;
        request(url, response => concatenateReadStream(response).then(resolve, reject));
      });
    }

    const queryIndex = url && url.lastIndexOf('?');
    url = queryIndex >= 0 ? url.substr(0, queryIndex) : url;
    const readFileAsync = util.promisify(fs.readFile);
    const buffer = yield readFileAsync(url, options);
    return buffer instanceof Buffer ? toArrayBuffer(buffer) : buffer;
  });
  return _readFile.apply(this, arguments);
}

function createReadStream(_x2, _x3) {
  return _createReadStream.apply(this, arguments);
}

function _createReadStream() {
  _createReadStream = _asyncToGenerator(function* (url, options) {
    if (isDataURL(url)) {
      return decodeDataUri(url);
    }

    if (!isRequestURL(url)) {
      return fs.createReadStream(url, options);
    }

    return new Promise((resolve, reject) => {
      options = _objectSpread({}, new URL(url), options);
      const request = url.startsWith('https:') ? https.request : http.request;
      request(url, response => resolve(response));
    });
  });
  return _createReadStream.apply(this, arguments);
}
